cps/elf.c

```diff
@@ -0,0 +1,406 @@
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	return string;
+}
+
+static struct string_table *
+strtbl_add(const char *str, struct string_table *tbl)
+{
+	if (tbl == NULL) {
+		unsigned int max = 64;
+		tbl = rte_malloc("strtbl",
+			sizeof(*tbl) + sizeof(char *) * max, 0);
+		if (tbl == NULL)
+			return NULL;
+		tbl->max = max;
+		tbl->cnt = 0;
+	}
+
+	if (tbl->cnt >= tbl->max) {
+		tbl->max *= 2;
+		tbl = rte_realloc(tbl,
+			sizeof(*tbl) + sizeof(char *) * tbl->max, 0);
+		if (tbl == NULL)
+			return NULL;
+	}
+
+	tbl->str[tbl->cnt] = str;
+	tbl->cnt += 1;
+
+	return tbl;
+}
+
+static struct string_table *
+PERBIT(load_strings)(struct elf_file *module, const char *secname,
+	struct string_table *tbl)
+{
+	unsigned long size;
+	const char *strings;
+
+	strings = PERBIT(get_section)(module, secname, &size);
+	if (strings == NULL)
+		return tbl;
+
+	if (strings[size - 1] != 0) {
+		CPS_LOG(WARNING,
+			"%s may be corrupt; an unterminated string was found at the end of section %s\n",
+			module->pathname, secname);
+	}
+
+	/* Skip any zero padding. */
+	while (!strings[0]) {
+		strings++;
+		if (size-- <= 1)
+			return tbl;
+	}
+
+	while (strings != NULL) {
+		tbl = strtbl_add(strings, tbl);
+		strings = next_string(strings, &size);
+	}
+
+	return tbl;
+}
+
+/*
+ * Check ELF file header.
+ */
+static int
+elf_ident(void *file, unsigned long fsize, int *conv)
+{
+	/* "\177ELF" <byte> where byte = 001 for 32-bit, 002 for 64. */
+	unsigned char *ident = file;
+
+	if (fsize < EI_CLASS || memcmp(file, ELFMAG, SELFMAG) != 0) {
+		/* Not an ELF object. */
+		return -ENOEXEC;
+	}
+
+	if (ident[EI_DATA] == 0 || ident[EI_DATA] > 2) {
+		/* Unknown endianness. */
+		return -EINVAL;
+	}
+
+	if (conv != NULL)
+		*conv = native_endianness() != ident[EI_DATA];
+
+	return ident[EI_CLASS];
+}
+
+/* Read ELF file into memory. */
+static struct elf_file *
+grab_elf_file(const char *pathname)
+{
+	struct elf_file *file;
+
+	file = rte_malloc("kni_elf_file", sizeof(*file), 0);
+	if (file == NULL)
+		goto fail;
+
+	file->pathname = rte_strdup("kni_elf_path", pathname);
+	if (file->pathname == NULL)
+		goto fail_free_file;
+
+	file->data = grab_file(pathname, &file->len);
+	if (file->data == NULL)
+		goto fail_free_pathname;
+
+	switch (elf_ident(file->data, file->len, &file->conv)) {
+	case ELFCLASS32:
+		/* FALLTHROUGH */
+	case ELFCLASS64:
+		break;
+	case -ENOEXEC:
+		/* FALLTHROUGH */
+	case -EINVAL:
+		/* FALLTHROUGH */
+	default:
+		goto fail_free_pathname;
+	}
+
+	return file;
+
+fail_free_pathname:
+	rte_free(file->pathname);
+fail_free_file:
+	rte_free(file);
+fail:
+	return NULL;
+}
+
+static void
+release_elf_file(struct elf_file *file)
+{
+	if (file == NULL)
+		return;
+
+	rte_free(file->data);
+	rte_free(file->pathname);
+	rte_free(file);
+}
+
+void
+print_kmod_info(const char *kmod_path)
+{
+	struct string_table *tags;
+	struct elf_file *mod;
+	unsigned int j;
+
+	mod = grab_elf_file(kmod_path);
+	if (mod == NULL) {
```

Comment:
You are obtaining the version from the file in the disk, not from the loaded module. Isn't it possible to obtain the version from the loaded module? The matter here is that print out the version information was motivated because if the file of the module was updated before running Gatekeeper, Gatekeeper won't update the loaded module, so printing out the version would let someone check what happened.

Reply:
It seems that you can read the srcversion field of the loaded kernel module from `/sys/modules/rte_kni/srcversion`, but the other fields (version, description, vermagic) are not held in the filesystem like srcversion.

When the module is already loaded, we could check the srcversion of the loaded kernel module, and if it doesn't match the srcversion of the .ko file that the user specified in configuration, then issue a warning or fail to start.

Reply:
Let's stick with field srcversion only and ignore the others.

That is a great solution since it makes the log entry silence (i.e. not logged) when there is no issue with the already loaded module.

Notice that according to the link below, the field srcversion is optional. So the code should log that it cannot check that the loaded the module is the same on disk due to the lack of the srcversion field in case it's already loaded.
https://stackoverflow.com/questions/3470081/modinfo-srcversion-how-do-i-generate-this-from-my-source/3474997